### PR TITLE
planner, workload-based learning: The setup process framework of workload-based learning

### LINKED ISSUE #58131: Table cost through workload-based learning
## Summary 

Table cost learning is the first workload based learning feature we build. It measures each table's contribution to workload cost, enabling tools to assess table significance relative to user workloads. For example, "auto analyze" can use this cost to rank tables in its priority queue. The write path is based on a async kernel process that analyze the workload (statement summary) and record the "cost" of each table used in the workload in a ew metadata table. The read path involves caching and updating the new metadata table and consuming the information by the priority queue system. 

## Catagory

TableCost is one kind of feedback feature in the whole workload-based learning picture

Image

## TableCost Overview

Producer part —— Tablecost worker
1. Load workload data from storage "statement summary table"
2. Analyze and compute the tablecost value
3. Write the tablecost into storage "workload values table"
Consumer part —— Feedback cache worker
1. Cache the feedback value —— table cost from storage into memory
2. Stats priority queue uses the table cost directly
Storage —— New Workload table
1. Used to save all of the workload values(table cost now) in this table 

Only data dependency between the producer part and the consumer part.

Image

### PR DESCRIPTION
**Pull request by elsa0520**
### What problem does this PR solve?

Issue Number: ref #58131

Problem Summary:

### What changed and how does it work?

The setup process framework of workload-based learning.
When tidb setup, the workload-based learnng worker such as read tablecost worker will setup at the same time.
It will start a new go routine to execute workers periodically. All workload-based learning related workers will be started in this function ```SetupWorkloadBasedLearningWorker"```

Two new configurations (TODO set value)
1. EnableWorkloadBasedLearning: Control whether enable all workload-based analyze workers. Default is false before this feature is ready
2. WorkloadBasedLearningInterval: Control the time interval between each round of analysis task. Default is 1 days.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Now, we can only disable it before the feature is not ready. So I manual test the setup worker.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by ghazalfamilyusa** [comment]
Do you actually mean that this PR close the corresponding 58131 issue?

**Review by ghazalfamilyusa — CHANGES_REQUESTED** [review]

File: pkg/domain/domain.go [hunk 1]
```diff
@@ -3329,6 +3343,38 @@ func (do *Domain) planCacheEvictTrigger() {
...
+	defer util.Recover(metrics.LabelDomain, "readTableCostWorker", nil, false)
+	readTableCostTicker := time.NewTicker(variable.WorkloadBasedLearningInterval.Load())
+	defer func() {
+		readTableCostTicker.Stop()
+		logutil.BgLogger().Info("readTableCostWorker exited.")
+	}()
+	for {
+		select {
+		case <-readTableCostTicker.C:
+			if variable.EnableWorkloadBasedLearning.Load() && do.statsOwner.IsOwner() {
```
Comment (qw4990) [inline pkg/domain/domain.go:3369]: Should we output any log for each execution?

File: pkg/workloadbasedlearning/handle.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workloadbasedlearning
+
+// Handle The entry point for all workload-based learning related tasks
+type Handle struct {
```
Comment (qw4990) [inline pkg/workloadbasedlearning/handle.go:18]: Better to add more comments to explain the background and motivation of this.

File: pkg/domain/domain.go [hunk 3]
```diff
@@ -3329,6 +3343,38 @@ func (do *Domain) planCacheEvictTrigger() {
 	}
 }
 
+// SetupWorkloadBasedLearningWorker sets up all of the workload based learning workers.
+func (do *Domain) SetupWorkloadBasedLearningWorker() {
+	wbLearningHandle := workloadbasedlearning.NewWorkloadBasedLearningHandle()
+	// Start the workload based learning worker to analyze the read workload by statement_summary.
+	do.wg.Run(
+		func() {
+			do.readTableCostWorker(wbLearningHandle)
```
Comment (qw4990) [inline pkg/domain/domain.go:3352]: If there are multiple TiDB instances in this cluster, can we guarantee there is only one node to execute this task?

File: pkg/domain/domain.go [hunk 4]
```diff
@@ -3329,6 +3343,38 @@ func (do *Domain) planCacheEvictTrigger() {
...
+		},
+		"readTableCostWorker",
+	)
+	// TODO: Add more workers for other workload based learning tasks.
+}
+
+func (do *Domain) readTableCostWorker(wbLearningHandle *workloadbasedlearning.Handle) {
```
[the remark this answers is not in the document]
Comment (elsa0520) [reply pkg/domain/domain.go:3359]: I added the comment where it is called. 
I will add the comment at here. But I think it is a litter duplicate comment.

File: pkg/domain/domain.go (hunk 1 above)
Comment (elsa0520) [reply pkg/domain/domain.go:3369]: I will output log inside of function "HandleReadTableCost"

File: pkg/domain/domain.go (hunk 3 above)
Comment (elsa0520) [reply pkg/domain/domain.go:3352]: Yes no problem

File: pkg/workloadbasedlearning/handle.go (hunk 2 above)
Comment (elsa0520) [reply pkg/workloadbasedlearning/handle.go:27]: Done

File: pkg/workloadbasedlearning/handle.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workloadbasedlearning
+
+/*
```
Comment (Rustin170506) [inline pkg/workloadbasedlearning/handle.go:17]: In Go, we rarely use `/* */` comments. See more at https://go.dev/doc/comment to check the package comment style.

**Review by Rustin170506 — APPROVED** [review]
Thanks! :shipit:

Some nits.

**Comment by elsa0520** [comment]
> Two points:
> 
> * Do you actually mean that this PR close the corresponding 58131 issue?
> * Need unit test for the initial functionality.

1. No, I changed the comment to ref 58131
2. I will add unit test in the future pr link. Only test the tidb setup is a little bit difficult in unit test

File: pkg/domain/domain.go [hunk 6]
```diff
@@ -3329,6 +3343,38 @@ func (do *Domain) planCacheEvictTrigger() {
...
+	defer util.Recover(metrics.LabelDomain, "readTableCostWorker", nil, false)
```
[the remark this answers is not in the document]
Comment (elsa0520) [reply pkg/domain/domain.go:3362]: DONE

File: pkg/workloadbasedlearning/handle.go (hunk 5 above)
Comment (elsa0520) [reply pkg/workloadbasedlearning/handle.go]: Done.

**Review by ghazalfamilyusa — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Review by benmeadowcroft — APPROVED** [review]

**Pull request merged**